### DISCUSSION TITLE
Anyone try self hosting hubs on a Raspberry Pi server?

### DISCUSSION THREAD

**Discussion by madjin** [mozilla/hubs#3099] [Hubs Cloud Questions] [answered]
Just got a Pi 4, anyone here try deploying Hubs Cloud onto a Raspberry Pi?

**Comment by netpro2k** [comment, marked as answer]
All the parts that make up Hubs are open source and you should in theory be able to get them running on most hardware. I have not tried a Pi4 in particular, but have set up most parts of the stack locally at one time or another on a few linux machines.it should be able to work from a technical perspective (not sure if you will run into any Arm specific issues on a Pi4).

That said, Hubs has a whole lot of moving parts, and getting everything configured and talking to each other can be pretty difficult, and we have not really tooled out things to make local deployment easy. It should be possible to stand each service up and get them all configured just so, but its going to take some digging on your part. The happy path may be trying to set things up using Chef habitat since that's how things are deployed to both hubs.mozilla.com and Hubs Cloud instances, but requires learning it and dealing with an extra layer of abstraction. You might instead try and individually stand up and configure each service independently, but this will require diving a bit deeper into the runtime requirements of each service (which span a variety of languages and frameworks).

If you do end up going down this path, we would be happy to try and answer questions, but just be aware you are treading into fairly uncharted waters. Sharing your results with the community would be super helpful, as its definitely something that comes up fairly often.

I also want to take this as an opportunity to explain more about why "Hubs Cloud" is built the way it is. We want Hubs to be an alternative to the typical walled garden you-dont-own-it-at-all model you see on other communication platforms. This means our  first focus has been to see if we can make self hosting as easy as signing up for a SaaS product. With the initial release of Hubs Cloud we are definitely not all the way there yet, but given the number and variety of people who have been able to successfully set up hubs cloud instances, I think we are very much on the right path. This does mean that we have had to add some complexity to the system to support the variety of usecases in which it might be deployed, and is much more "productionized" than a comparatively simple "just throw it on a box" setup might be. This is not to say that running Hubs on your own hardware is a non-goal, its just a matter of where we prioritize our efforts.
